### PR TITLE
Add Debug impl for NodeEvent

### PR DESCRIPTION
**Pull request by dvdplm**
**NOTE**: adds a `Debug` bound on `TUserData` which might not be desired

Builds on https://github.com/libp2p/rust-libp2p/pull/535

### CODE REVIEW

**Comment by tomaka** [comment]
I don't think we should test the `Debug` trait. It's output is not guaranteed.

**Comment by debris** [comment]
why is it not guaranteed?

**Comment by dvdplm** [comment]
Sorry for a poor PR descr; I wanted to ask about the `todo` [here](https://github.com/libp2p/rust-libp2p/pull/538/commits/c9ec8ef5e42c96f702856db3a68b9bbe23ca954a#diff-25c5a42eda032f9031093ae96bccc789L331) where the commented out code must have been for an older version. I'm honestly not sure if there's any real value in impl-ing `Debug` manually as opposed to just `#[derive(Debug)]`. Maybe you had something different in mind for the `TODO`?

As for testing `Debug`, I don't have a strong opinion. In my limited experience it can be useful sometimes; a broken debug output can signal something else. It's also bound to be pretty brittle (but trivial to fix when it breaks). @tomaka: Your call! :)

**Comment by tomaka** [comment]
> why is it not guaranteed?

That's just my opinion, but `Debug` is just for displaying information to the programmer for debugging purposes, so to me changing what is displayed shouldn't break anything.

**Comment by dvdplm** [comment]
@tomaka We could keep the `Debug` impl and ditch the test if you prefer it (or move the test to its own PR).

**Comment by tomaka** [comment]
> NOTE: adds a Debug bound on TUserData which might not be desired

The reason why I added a separate implementation of `fmt::Debug` rather than deriving Debug is to avoid requiring `Debug` on `TUserData`.
It's debatable whether this is a good decision, but if you're requiring `Debug` on `TUserData` then you might as well just add `derive(Debug)`.

File: core/src/nodes/node.rs [hunk 1]
```diff
@@ -326,32 +325,39 @@ where
...
+    TMuxer: muxing::StreamMuxer,
+    TUserData: fmt::Debug,
 {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
         match self {
-            NodeEvent::Incoming { ref listen_addr, .. } => {
-                f.debug_struct("NodeEvent::Incoming")
-                    .field("listen_addr", listen_addr)
+            NodeEvent::Multiaddr(result) => {
+                write!(f, "{:?}", result)
```
Comment (tomaka) [inline core/src/nodes/node.rs]: This should indicate that it's an event, not only print the multiaddr.

**Comment by twittner** [comment]
Adding my two cents here: I would just derive `Debug` for `NodeEvent` and be done with it. Logging event data is common enough that having `TUserData` implement `Debug` should not be too much to ask for, at least not if the expectation is that `NodeEvent` implements `Debug`.

**Comment by dvdplm** [comment]
I'm not clear on what the decision is here, but I think we want to close this PR and https://github.com/libp2p/rust-libp2p/pull/562 as well and instead just `#[derive(Debug)]` on `NodeEvent`. 
Is that correct?

**Pull request closed without merging**